# Exploring Ebay Car Sale Data

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('data_source/autos.csv', encoding='latin-1')

In [2]:
autos #Present the first few and last few data from the csv file

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

From the obsevation of the data list, there are a total of 50,000 entries of data with 20 columns, making up of objects and integer. Carefully looking through the columns, are are missing data found some columns such as 'vehicleType, gearbox, model, fuelType, and notRepairedDamage'.

Datetime, price, odometer, will need to be clean for easy analysis.

In [4]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


## Cleaning the index title to snakecase

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'
] #All columns value must be included using this method to be error free

autos.head()

#Tried using .map but it doesn't work. "TypeError: 'Index' object is not callable"
#.map can't be use to change content of index. 
#Items not included in .map will become nan


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Cleaning index title, solution 2:

In [6]:
'''
import re

def cleaned_cols(col):
    col = col.replace('yearOfRegistration','registration_year')
    col = col.replace('monthOfRegistration','registration_month')
    col = col.replace('notRepairedDamage','unrepaired_damage')
    col = col.replace('dateCreated','ad_created')
    # Change the rest of the column names from camelcase to snakecase
    col = re.sub( '(?<!^)(?=[A-Z])', '_', col ).lower()
    return col

autos.columns = [cleaned_cols(c) for c in autos.columns]

autos.columns
'''

"\nimport re\n\ndef cleaned_cols(col):\n    col = col.replace('yearOfRegistration','registration_year')\n    col = col.replace('monthOfRegistration','registration_month')\n    col = col.replace('notRepairedDamage','unrepaired_damage')\n    col = col.replace('dateCreated','ad_created')\n    # Change the rest of the column names from camelcase to snakecase\n    col = re.sub( '(?<!^)(?=[A-Z])', '_', col ).lower()\n    return col\n\nautos.columns = [cleaned_cols(c) for c in autos.columns]\n\nautos.columns\n"

Cleaning index title, solution 3:

In [7]:
#Targeted changes just like solution 2 using .rename.
#pd.rename({'ram': 'ram_gb','storage':'stor-age'}, axis=1, inplace=True)

## Determine what other cleaning tasks need to be done 

In [8]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

- Price and odometer value need to be cleaned to make it numerical values.
- nr_of_pictures has no value.

In [9]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(float)

In [10]:
#Changes name of odometer for easy ref for future
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

autos['odometer_km'] = autos['odometer_km'].str.replace('km','')
autos['odometer_km'] = autos['odometer_km'].str.replace(',','')
autos['odometer_km'] = autos['odometer_km'].astype(float)

## Continue to analyze odometer_km and price columns that doesn't look right 

Look for an outliers that might need to be remove.

In [11]:
autos['odometer_km'].unique().shape[0]

13

In [12]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [13]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

In [14]:
autos['price'].unique().shape[0]

2357

In [15]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos['price'].value_counts().sort_index(ascending=False).head(15)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64

Through further analysis on price data, we observe 1421 entries are $0 which seems unlikely. As it is made up of 2.8% of the data presented in price column, we should be able safely remove those rows. 

Code to remove outliers: df[df["col"].between(x,y)]

In [17]:
print('mean:')
print(autos['price'].mean())
print('median:')
print(autos['price'].median())
print('mode:')
print(autos['price'].mode())

mean:
9840.04376
median:
2950.0
mode:
0    0.0
dtype: float64


Looking at the mean, median and mode of price column, median and mode suggesting that the distribution curve is positively skew, although, mean shows the other way. This could mean that the prices are so high that it distort the mean value.

## Looking into date time columns 

As date_crawled, last_seen, ans ad_created columns are represented as strings, we need to convert the data into numerical representation so we can understand it quantitatively.

In [18]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [19]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

The crawler craw the data consistently.

In [20]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-01-22    0.00002
2016-01-27    0.00006
2016-01-29    0.00002
2016-02-01    0.00002
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-07    0.00002
2016-02-08    0.00002
2016-02-09    0.00004
2016-02-11    0.00002
2016-02-12    0.00006
2016-02-14    0.00004
2016-02-16    0.00002
2016-02-17    0.00002
2016-02-18    0.00004
2016-02-19    0.00006
2016-02-20    0.00004
2016-02-21    0.00006
               ...   
2016-03-09    0.03324
2016-03-10    0.03186
2016-03-11    0.03278
2016-03-12    0.03662
2016-03-13    0.01692
2016-03-14    0.03522
2016-03-15    0.03374
2016-03-16    0.03000
2016-03-17    0.03120
2016-03-18    0.01372
2016-03-19    0.03384
2016-03-20    0.03786
2016-03-21    0.03772
2016-03-22    0.03280
2016-03-23

From the obervation on ad_created, more ads are created from march 2016 onwards in comparision to earlier dates.

In [21]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [22]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Through the descriptive analysis above, we observe that the min year is 1000 and max 9999 which is illogical. As such, we will adjust the years to somewhat more realistic when cars are invented.

In [23]:
autos[autos['registration_year'].between(1900,2016)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [24]:
autos['registration_year'].value_counts(normalize=True, ascending=False)

2000    0.06708
2005    0.06030
1999    0.06000
2004    0.05474
2003    0.05454
2006    0.05416
2001    0.05406
2002    0.05066
1998    0.04906
2007    0.04608
2008    0.04462
2009    0.04196
1997    0.04056
2011    0.03268
2010    0.03194
2017    0.02906
1996    0.02888
2012    0.02646
2016    0.02632
1995    0.02626
2013    0.01612
2014    0.01332
1994    0.01320
2018    0.00984
1993    0.00890
2015    0.00798
1990    0.00790
1992    0.00782
1991    0.00712
1989    0.00362
         ...   
1950    0.00006
1955    0.00004
9000    0.00004
1954    0.00004
1800    0.00004
1957    0.00004
1941    0.00004
1951    0.00004
1934    0.00004
4100    0.00002
4800    0.00002
1953    0.00002
1111    0.00002
1927    0.00002
6200    0.00002
4500    0.00002
1943    0.00002
5911    0.00002
1939    0.00002
1938    0.00002
2800    0.00002
8888    0.00002
1000    0.00002
1500    0.00002
1948    0.00002
1931    0.00002
1929    0.00002
1001    0.00002
9996    0.00002
1952    0.00002
Name: registration_year,

In [25]:
autos['registration_year'].between(1900,2016).sum() / autos.shape[0]

0.96056

Registration year from year 1900 to 2016 accounts for 96%.

## Understanding the most common car brand on Ebay

In [26]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

From the result above, we can identify that the common car that take up a big portion are VW, opel, bmw. Our further analysis will only focus on the top 5% of the car brand.

In [27]:
brands = autos['brand'].value_counts(normalize=True)
top_5_precent = brands[brands > 0.05].index #unique
top_5_precent

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [28]:
brand_and_prices = {}

for c in top_5_precent:
    #Loop the unique sector with the data
    car_brand = autos[c == autos['brand']]
    #Find the average price of car from each brand
    avg_price = car_brand.sort_values('price', ascending = False).mean().iloc[0]
    brand_and_prices[c] = int(avg_price)
    
brand_and_prices

{'volkswagen': 6384,
 'opel': 5106,
 'bmw': 8252,
 'mercedes_benz': 29511,
 'audi': 8965,
 'ford': 7105}

volkswagen        0.21374 = 21%, avg price = 6384 |
opel              0.10922 = 10%, avg price = 5106 |
bmw               0.10858 = 10%, avg price = 8252 |
mercedes_benz     0.09468 = 9%, avg price = 29511 |
audi              0.08566 = 8%, avg price = 8965 |
ford              0.06958 = 6%, avg price = 7105 |
 
After calculating the avg price of each car brand from the top 5%, we can see that VW is one of the lowest in price car other than opel, as such, the data suggest that the price play a big part on consumer decision of their car purchase. 

Though further observation, we can see that consumer either purchase an affordable car or a popular branded car. 

In [29]:
brand_mean_mileage = {}

for brands in top_5_precent:
    #Loop the unique sector with the data
    brand_only = autos[autos['brand'] == brands]
    #Find the avg mileage from each brand
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brands] = int(mean_mileage)
    
brand_mean_mileage

{'volkswagen': 128955,
 'opel': 129298,
 'bmw': 132521,
 'mercedes_benz': 130886,
 'audi': 129643,
 'ford': 124131}

In [30]:
#Convert dictionary into Series
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending = False)
mean_price = pd.Series(brand_and_prices).sort_values(ascending = False)
mean_mileage

bmw              132521
mercedes_benz    130886
audi             129643
opel             129298
volkswagen       128955
ford             124131
dtype: int64

In [31]:
#Convert series into dataframe
top_brand_info = pd.DataFrame(mean_mileage, columns = ['mean_mileage'])
top_brand_info

mean_mileage
bmw                  132521
mercedes_benz        130886
audi                 129643
opel                 129298
volkswagen           128955
ford                 124131

In [32]:
#Add another columns to the table
top_brand_info['mean_price'] = mean_price
top_brand_info

mean_mileage  mean_price
bmw                  132521        8252
mercedes_benz        130886       29511
audi                 129643        8965
opel                 129298        5106
volkswagen           128955        6384
ford                 124131        7105

The mean mileage and price of the cars does not sure a distinction if there is a correlation between the 2 variable.

In [33]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The following task is to change all the dates "2016-03-21" to uniform numeric date 20160321

In [34]:
def clean_date(col):
    col = col.str.replace('-', '')
    return col

autos['date_crawled'] = clean_date(autos['date_crawled'])
autos['ad_created'] = clean_date(autos['ad_created'])
autos['last_seen'] = clean_date(autos['last_seen'])

Next step is to identify all german words and translate them in english words.

In [35]:
autos['gearbox'].value_counts()

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

In [36]:
autos['unrepaired_damage'].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

In [37]:
autos['fuel_type'].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [38]:
autos['vehicle_type'].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [39]:
gearbox = {
    'manuell': 'manual',
    'automatik': 'automatic'
}

unrepaired_damage = {
    'nein': 'no',
    'ja': 'yes'
}

autos['gearbox'] = autos['gearbox'].map(gearbox)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage)
autos['fuel_type'] = autos['fuel_type'].replace(['benzin', 'elektro', 'andere'], ['gasoline', 'electric', 'other'])
autos['vehicle_type'] = autos['vehicle_type'].replace(['limousine', 'kleinwagen', 'kombi', 'cabrio', 'andere'], ['sedan', 'small car', 'stationwagen', 'convertible', 'other'])
autos['model'] = autos['model'].replace('andere', 'other')
autos['seller'] = autos['seller'].replace(['privat','gewerblich'],['Private', 'commercial'])

autos.head()

date_crawled                                               name  \
0  20160326 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  20160404 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  20160326 18:57:24                         Volkswagen_Golf_1.6_United   
3  20160312 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  20160401 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller offer_type   price   abtest  vehicle_type  registration_year  \
0  Private    Angebot  5000.0  control           bus               2004   
1  Private    Angebot  8500.0  control         sedan               1997   
2  Private    Angebot  8990.0     test         sedan               2009   
3  Private    Angebot  4350.0  control     small car               2007   
4  Private    Angebot  1350.0     test  stationwagen               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0     manual       158   other     150000.0                   3       lpg   
1  automatic       286     7er     150000.0                   6  gasoline   
2     manual       102    golf      70000.0                   7  gasoline   
3  automatic        71  fortwo      70000.0                   6  gasoline   
4     manual         0   focus     150000.0                   7  gasoline   

        brand unrepaired_damage         ad_created  nr_of_pictures  \
0     peugeot                no  20160326 00:00:00               0   
1         bmw                no  20160404 00:00:00               0   
2  volkswagen                no  20160326 00:00:00               0   
3       smart                no  20160312 00:00:00               0   
4        ford                no  20160401 00:00:00               0   

   postal_code          last_seen  
0        79588  20160406 06:45:54  
1        71034  20160406 14:45:08  
2        35394  20160406 20:15:37  
3        33729  20160315 03:16:28  
4        39218  20160401 14:38:50

How much cheaper are cars with damage than their non-damaged coutnerparts?

In [40]:
autos['unrepaired_damage'].unique()

array(['no', nan, 'yes'], dtype=object)

In [41]:
damages = autos['unrepaired_damage'].unique()

damaged_prices = {}

for d in damages:
    damage = autos[autos['unrepaired_damage'] == d]
    mean_price = damage['price'].mean()
    damaged_prices[d] = mean_price
    
damaged_prices

{'no': 8470.59579359673, nan: nan, 'yes': 2311.889248835797}

We can see that no damage cars are more expensive than damage cars.

Split odometer_km into groups:

In [42]:
print(autos['odometer_km'].value_counts().sort_index(ascending=True))

5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64


Does average prices follows any pattern based on milage?

In [43]:
milage_and_price = {}

odometer = autos['odometer_km'].unique()
for o in odometer:
    #Loop the unique milage with the data
    unique_odometer = autos[autos['odometer_km'] == o]
    #Find the avg price for each category of milage
    avg_price = unique_odometer.sort_values('price', ascending = False).mean().iloc[0]
    milage_and_price[o] = int(avg_price)
    
milage_and_price

{150000.0: 8280,
 70000.0: 10753,
 50000.0: 25596,
 80000.0: 9473,
 10000.0: 19404,
 30000.0: 16343,
 125000.0: 6197,
 90000.0: 8259,
 20000.0: 17547,
 60000.0: 12148,
 5000.0: 11397,
 100000.0: 12428,
 40000.0: 48765}

In [44]:
mean_milage_price = pd.Series(milage_and_price).sort_values(ascending = False)

In [45]:
#Convert series into dataframe
table_milage_price = pd.DataFrame(mean_milage_price, columns = ['mean_milage_price'])
table_milage_price

mean_milage_price
40000.0               48765
50000.0               25596
10000.0               19404
20000.0               17547
30000.0               16343
100000.0              12428
60000.0               12148
5000.0                11397
70000.0               10753
80000.0                9473
150000.0               8280
90000.0                8259
125000.0               6197

In [46]:
table_milage_price['unique_milage_post'] = autos['odometer_km'].value_counts().sort_index(ascending=True)
table_milage_price

#table_milage_price.drop('unique_milage', axis=1)

mean_milage_price  unique_milage_post
40000.0               48765                 819
50000.0               25596                1027
10000.0               19404                 264
20000.0               17547                 784
30000.0               16343                 789
100000.0              12428                2169
60000.0               12148                1164
5000.0                11397                 967
70000.0               10753                1230
80000.0                9473                1436
150000.0               8280               32424
90000.0                8259                1757
125000.0               6197                5170

Through the observation from the table above, we can see that cars in the milage of 40000 and 50000 hold a high prices compare with the rest. 

1 hypothesis is that the price depend on milage and the supply and demand on Ebay. The higher the supply the lower the price which can be seen on 150,000 and 125,000 milage with 32,424 and 5170 posts respectively. In comparision to higher price with lower milage and posts which can be seen on the first 5 of the table. 

1 anomaly is the 5000 milage with less than a thousand posts but cost alot lower than those milage on the top 5 listed on the table.